# **Tarea Clase Asincrónica - Ejercicio sobre: ¿Qué socio del gimnasio va a cancelar?**


**Parte A**· Entender el problema (Unidad 1)

1.	**¿Es regresión o clasificación? ¿Por qué?**

**Respuesta/:** Es un ejercicio de Clasificación. ya que se trata de predecir una categoria (Si o No) y no un valor númerico.

2.	**Identifica X y Y.**

**X:** Datos que nos ayuden a realizar la predicción (meses_activo, visitas_mes, quejas, plan)

**Y:** Variable que queremos predecir: SI o NO de la columna cancelo.

3. **Pregunta de reflexión: si agregáramos la columna «recibió encuesta de salida», ¿sería buena idea usarla como X? Explica (pista: información «del futuro»).**

Al ser información que se obtiene **al final**, una vez que la persona h**a decidido cancelar**, no es buena idea ya que no es un variable que nos ayude a realizar la predicción.


In [8]:
#Parte B · Preparar los datos (Unidad 2)

#1.	Carga los datos y cuenta cuántos socios cancelaron y cuántos no

#Importar librerias
import pandas as pd #analizar tablas de datos
from sklearn.model_selection import train_test_split #Dividir datos en entrenamiento y prueba
from sklearn.linear_model import LogisticRegression # algoritmo de clasificación para Regresión Logística
from sklearn.tree import DecisionTreeClassifier # algoritmo de clasificación para el árbol de decisión
from sklearn.metrics import (accuracy_score, precision_score, recall_score, confusion_matrix) #métricas para evaluar el modelo

#Cargar los datos desde el archivo.csv
df = pd.read_csv('gimnasio_cancelaciones.csv')  # lee el archivo subido al colab y lo guarda en df

#Contar cuantos socios cancelaron y cuantos no - de la columna cancelo
print ("Conteo de Cancelaciones: ")
print(df['cancelo'].value_counts())  # muestra el total de si y no

# Convertir el texto en números:
#	plan_mensual: 1 si el plan es «Mensual», 0 si es «Anual».
#	cancelo: 1 si es «Sí», 0 si es «No».
df['plan'] = df['plan'].map({"Mensual": 1, "Anual": 0})
df['cancelo'] = df['cancelo'].map({"Sí": 1, "No": 0})

print(df[["plan", "cancelo"]].head())

# Separa en entrenamiento y prueba con test_size=0.3 esto deja el 30% de los datos para evaluar el modelo
# y stratify=y : mantiene la misma proporcion de cancelados en ambos grupos
#random_state=42 : hace uso de los mismos datos cada ve que se ejecute
#(para que ambos grupos tengan la misma proporción de cancelaciones).

X = df.drop(  # agarra todo el dataset quitando la columna cancelo
    columns=['cancelo']
)

y = df['cancelo']  #  Y se quda con la columna objetivo: cancelo

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, stratify=y, random_state=42
)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, stratify=y, random_state=42
)


Conteo de Cancelaciones: 
cancelo
No    195
Sí    105
Name: count, dtype: int64


In [16]:
#Parte C · Entrenar tres modelos (Unidad 3)


# MODELO 1: Regresión Logística:
modelo_log = (  # inicializa la regresión logistica de sklearn
    LogisticRegression()
)

modelo_log.fit (   # entrena el modelo con el set de datos X y Y de entrenamiento
    X_train, y_train)

#MODELO 2: Árbol sin límite: no le ponemos restricción de profundidad
modelo_arbol_libre = (
    DecisionTreeClassifier(random_state=42)
)

modelo_arbol_libre.fit (
    X_train, y_train)

# MODELO 3: Árbol Limitado: limitamos el crecimiento del arbol a max_depth =3 para evitar sobreajuste (overfitting)
modelo_arbol_limite = (
    DecisionTreeClassifier(max_depth=3, random_state=42)
)

modelo_arbol_limite.fit (
    X_train, y_train)

#Evaluar los tres modelos:

#Hacemos las predicciones para el grupo de prueba con cada modelo
pred_log = modelo_log.predict(X_test)
pred_arbol_libre = modelo_arbol_libre.predict(X_test)
pred_arbol_limite = modelo_arbol_limite.predict(X_test)

#Mostramos los resultados:
print(" \n *** Regresión Logística: ***")
print(f"Exactitud (Accuracy): {accuracy_score(y_test, pred_log):.2f}")
print(f"Exhaustividad (Recall): {recall_score(y_test, pred_log):.2f}")
print(f"Precisión (Precision): {precision_score(y_test, pred_log):.2f}")

print("\n *** Árbol sin límite: ***")
print(f"Exactitud (Accuracy): {accuracy_score(y_test, pred_arbol_libre):.2f}")
print(f"Exhaustividad (Recall): {recall_score(y_test, pred_arbol_libre):.2f}")
print(f"Precisión (Precision): {precision_score(y_test, pred_arbol_libre):.2f}")

print("\n *** Árbol limitado: ***")
print(f"Exactitud (Accuracy): {accuracy_score(y_test, pred_arbol_limite):.2f}")
print(f"Exhaustividad (Recall): {recall_score(y_test, pred_arbol_limite):.2f}")
print(f"Precisión (Precision): {precision_score(y_test, pred_arbol_limite):.2f}")

 
 *** Regresión Logística: ***
Exactitud (Accuracy): 0.93
Exhaustividad (Recall): 0.90
Precisión (Precision): 0.90

 *** Árbol sin límite: ***
Exactitud (Accuracy): 0.80
Exhaustividad (Recall): 0.71
Precisión (Precision): 0.71

 *** Árbol limitado: ***
Exactitud (Accuracy): 0.84
Exhaustividad (Recall): 0.77
Precisión (Precision): 0.77


**•	Una conclusión final de 3 a 5 líneas: ¿qué modelo recomiendas y por qué?**

Según los resultados que se obtuvieron, se puede determinar que utilizar el modelo de Regresión Logítica sería lo ideal, ya que obtuvo mejores resultados de desempeño: Exactitud del 93% y un Recall de 90% al igual que precisión en los datos de prueba. por lo que se concluye que, se puede detectar que en el 90% de los socios, se puede detectar quienes tienen la intención de cancelar su plan adquirido